[BE][DDP] enable rebuilt bucket when find_unused_parameters=True

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73276

There are two major cases when find_unused_parameters=True:
1. grad ready order does not change over iterations, in this case, enable rebuilt bucket after first iteration can potentially improve performance
2. grad ready order changes over iterations, in this case, use static bucket order or dynamic bucket order in the first iteration does not matter much, as order changes per iteration

Differential Revision: [D34410523](https://our.internmc.facebook.com/intern/diff/D34410523/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34410523/)!

zhaojuanmao committed:
enable rebuilt bucket when find_unused_parameters=True

There are two major cases when find_unused_parameters=True:
1. grad ready order does not change over iterations, in this case, enable rebuilt bucket after first iteration can potentially improve performance
2. grad ready order changes over iterations, in this case, use static bucket order or dynamic bucket order in the first iteration does not matter much, as order changes per iteration

Differential Revision: [D34410523](https://our.internmc.facebook.com/intern/diff/D34410523/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34410523/)!

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -192,10 +184,9 @@ class TORCH_API Reducer {
+  // buckets once after the first iteration.
   inline bool should_rebuild_buckets() const {
-    return (static_graph_ || !find_unused_parameters_) && !has_rebuilt_bucket_;
+    return !has_rebuilt_bucket_;
```
Curious, should we add a knob (say, env var) to disable this feature just for debugging purposes?

zhaojuanmao replied:
> Do we want to eventually consider rebuilding buckets again when the graph has changed when training dynamic graph? Depending on the use case and how common it is this could also have a perf improvement.
> 
> Also, I'm assuming that this change will cause unused parameters to be allreduced first, because we mark unused parameters immediately when the first hook fires. Is this best for performance or should unused parameters be the last to allreduce?

rebuilt bucket every iteration for dynamic graph will be expensive, as rebuilt bucket needs a broadcast call. I added a comment in the codes.  

in line 836, we pushed the unused parameters at the end of buckets for better performance

zhaojuanmao replied:
> Curious, should we add a knob (say, env var) to disable this feature just for debugging purposes?

sounds good to me

mrshenli commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -192,10 +184,9 @@ class TORCH_API Reducer {
-  // find_unused_parameters_.
+  // buckets once after the first iteration.
   inline bool should_rebuild_buckets() const {
-    return (static_graph_ || !find_unused_parameters_) && !has_rebuilt_bucket_;
```
Could you please add the PR summary's descriptions as comments here as well? thanks!

zhaojuanmao replied:
sounds good!

zhaojuanmao committed:
Update on "enable rebuilt bucket when find_unused_parameters=True"


There are two major cases when find_unused_parameters=True:
1. grad ready order does not change over iterations, in this case, enable rebuilt bucket after first iteration can potentially improve performance
2. grad ready order changes over iterations, in this case, use static bucket order or dynamic bucket order in the first iteration does not matter much, as order changes per iteration

Differential Revision: [D34410523](https://our.internmc.facebook.com/intern/diff/D34410523/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34410523/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "enable rebuilt bucket when find_unused_parameters=True"


There are two major cases when find_unused_parameters=True:
1. grad ready order does not change over iterations, in this case, enable rebuilt bucket after first iteration can potentially improve performance
2. grad ready order changes over iterations, in this case, use static bucket order or dynamic bucket order in the first iteration does not matter much, as order changes per iteration

Differential Revision: [D34410523](https://our.internmc.facebook.com/intern/diff/D34410523/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34410523/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.hpp:
```diff
@@ -175,27 +181,34 @@ class TORCH_API Reducer {
+  // if users want to disable this feature for debugging purpose.
   inline bool should_rebuild_buckets() const {
-    return (static_graph_ || !find_unused_parameters_) && !has_rebuilt_bucket_;
+    return (static_graph_ || !find_unused_parameters_ ||
```
@rohan-varma @zhaojuanmao PR title says "enable rebuilt bucket when find_unused_parameters=True", however this method will return False if find_unused_parameters_ = True and static_graph_=False. Am I missing something here?

zhaojuanmao replied:
the below added one more condition (|| parse_env("DISABLE_REBUILT_BUCKET").compare("1") != 0), when static_graph=False && find_unused_parameters_=True && DISABLE_REBUILT_BUCKET!=1, it will return True
